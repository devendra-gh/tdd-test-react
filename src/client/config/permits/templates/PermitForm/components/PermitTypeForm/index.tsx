import React from 'react';
import { IVariables } from '@tamm/app-composer';
import Form from '@tamm/ui-lib-v2-form';
import { checkValidationField } from 'client/config/utils/checkValidation';
import CustomFormGroup from '../CustomGroup';
import '../../PermitForm.less';

const PermitTypeForm = (props: IVariables) => {
  const {
    values,
    handleChange,
    getFormFields,
    i18n,
    startShowingErrors,
    serviceType,
    onSubmit,
  } = props;

  if (!serviceType) {
    return null;
  }

  const formFields: IVariables = getFormFields() || {};
  const FormFields = Object.values(formFields).map((field: IVariables) => {
    return (
      <CustomFormGroup
        inputGroups={[field]}
        handleChange={handleChange}
        values={values}
        i18n={i18n}
        key={`${serviceType}.${field.stateKey}`}
        groupKey={`${serviceType}.${field.stateKey}`}
        startShowingErrors={startShowingErrors}
        validate={(fieldConfig: IVariables, stateKey: string) => {
          return fieldConfig.validationConfig
            ? checkValidationField(
                fieldConfig.validationConfig,
                values[stateKey][fieldConfig.name] || null,
                startShowingErrors,
                i18n,
              )
            : undefined;
        }}
        customDisabledDate={(currentDate: string, fieldDate: IVariables) =>
          typeof fieldDate.customDisabledDate === 'function' &&
          fieldDate.customDisabledDate(currentDate, values)
        }
      />
    );
  });

  return (
    <React.Fragment>
      {FormFields}
      <Form
        id="form-buttons"
        submitButton={{
          label: i18n('submit'),
          // withArrow: true,
          onClick: onSubmit,
        }}
        cancelLink={{
          label: i18n('cancel'),
          href: '#',
          onClick: () => {
            window.location.href = `${
              window.location.href.indexOf('stage.tamm') !== -1
                ? 'https://stage.tamm.abudhabi/'
                : 'https://www.tamm.abudhabi/'
            }tamm-centers-services/department-of-economic-development`;
          },
        }}
      />
    </React.Fragment>
  );
};

export default PermitTypeForm;
