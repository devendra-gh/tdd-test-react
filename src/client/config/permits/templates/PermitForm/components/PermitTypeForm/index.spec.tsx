import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import PermitTypeForm from './index';

jest.mock('client/config/utils/checkValidation', () => ({
  checkValidationField: jest.fn(),
}));
jest.mock('client/utils/baseUrl', jest.fn());

// eslint-disable-next-line no-console
console.log = () => {};
console.info = () => {};
console.warn = () => {};
console.error = () => {};

describe('Permit Type form', () => {
  let props: any;

  beforeEach(() => {
    props = {
      i18n: (value: string) => value,
      serviceType: null,
      onSubmit: jest.fn(),
      values: {
        testStateKey: {},
      },
      validation: {},
      actions: {
        validation: {
          update: jest.fn(),
        },
      },
      getFormFields: () => ({
        testField: {
          name: 'test field',
          stateKey: 'testStateKey',
          fields: [
            {
              validationConfig: {},
              elementType: 'input',
              onChange: () => jest.fn(),
            },
          ],
        },
      }),
    };
  });

  it('should successfully render when serviceType is provided', () => {
    props.serviceType = 'testServiceType';
    const component = shallow(<PermitTypeForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should not render when serviceType is falsy', () => {
    const component = shallow(<PermitTypeForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('getFormFields is null', () => {
    props.serviceType = 'testServiceType';
    props.getFormFields = () => null;
    const component = mount(
      <Router>
        <PermitTypeForm {...props} />
      </Router>,
    );
    const submitButton = component.find('Button');
    submitButton.simulate('click');
    expect(props.onSubmit).toHaveBeenCalled();
  });

  it('should submit form when validation passes', () => {
    props.serviceType = 'testServiceType';
    const component = mount(
      <Router>
        <PermitTypeForm {...props} />
      </Router>,
    );
    const submitButton = component.find('Button');
    submitButton.simulate('click');
    expect(props.onSubmit).toHaveBeenCalled();
  });

  it('should run date validation for date fields', () => {
    props.serviceType = 'testServiceType';
    const component = shallow(<PermitTypeForm {...props} />);
    const CustomFormGroup = component.find('CustomFormGroup');
    const fieldDate = {
      customDisabledDate: jest.fn(),
    };
    const formGroupProps: any = CustomFormGroup.props();
    const { customDisabledDate } = formGroupProps;
    customDisabledDate(new Date(), fieldDate);
    expect(fieldDate.customDisabledDate).toHaveBeenCalled();
  });
});
