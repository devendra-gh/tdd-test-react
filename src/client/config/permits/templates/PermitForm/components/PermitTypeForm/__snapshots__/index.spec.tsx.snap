// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permit Type form should not render when serviceType is falsy 1`] = `""`;

exports[`Permit Type form should successfully render when serviceType is provided 1`] = `
<Fragment>
  <CustomFormGroup
    customDisabledDate={[Function]}
    groupKey="testServiceType.testStateKey"
    i18n={[Function]}
    inputGroups={
      Array [
        Object {
          "fields": Array [
            Object {
              "elementType": "input",
              "onChange": [Function],
              "validationConfig": Object {},
            },
          ],
          "name": "test field",
          "stateKey": "testStateKey",
        },
      ]
    }
    key="testServiceType.testStateKey"
    validate={[Function]}
    values={
      Object {
        "testStateKey": Object {},
      }
    }
  />
  <LocalizedComponent(Form)
    backButton={null}
    cancelLink={
      Object {
        "href": "#",
        "label": "cancel",
        "onClick": [Function],
      }
    }
    help=""
    i18n={[Function]}
    id="form-buttons"
    submitButton={
      Object {
        "label": "submit",
        "onClick": [MockFunction],
      }
    }
    validateStatus={null}
  />
</Fragment>
`;
